### PR TITLE
aggregator: unify resolver implementation and tests

### PR DESCRIPTION
**Pull request by sttts**
This is https://github.com/kubernetes/kubernetes/pull/45082, but without the port support.

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 1]
```diff
@@ -29,51 +29,98 @@ import (
...
+}
+
+// resolveToService returns the service, scheme and the port string (can be a name, can be a numeric port).
+func resolveToService(services listersv1.ServiceLister, namespace, id string) (*v1.Service, string, string, error) {
+	svc, err := services.Services(namespace).Get(id)
+	if err != nil {
+		return nil, "", "", err
 	}
 
+	return svc, "https", "443", nil
```
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: These seem like good defaults but if the service object already has the port and scheme, then either grab those and return them or just return the service and leave it to the caller to worry about extracting/determining defaults. (We already seem to handle the port portion of this in your findServicePort method) This also suggests we are dropping support for "HTTP" when talking to aggregated api servers. I think thats probably a good thing but I want to make sure its a deliberate choice.

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 2]
```diff
@@ -29,51 +29,98 @@ import (
...
+
 	// Find a Subset that has the port.
 	for ssi := 0; ssi < len(eps.Subsets); ssi++ {
 		ss := &eps.Subsets[(ssSeed+ssi)%len(eps.Subsets)]
 		if len(ss.Addresses) == 0 {
 			continue
 		}
 		for i := range ss.Ports {
-			if ss.Ports[i].Name == portStr {
+			if ss.Ports[i].Name == svcPort.Name {
```
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: Not for now but there is something odd about the case where we are asked for a port number match, find the first matching port number and the pick all the routes which match that routes port name. If ports are consistently names there won't be an issue but if they aren't the behavior is going to be very hard to debug.

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go (hunk 1 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: Updated the code here from #45082, i.e. the caller decides.

The HTTP removal is intentional, as discussed in #45082 with @deads2k. For dev and testing one can disable cert verification and use self-signed certs. So we don't need HTTP anymore.

**Comment by sttts** [comment]
@cheftako addressed your comments. Also removed the remaining forbidden import in the tests. Ready for final review.

**Comment by deads2k** [comment]
lgtm
/approve

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 3]
```diff
@@ -24,61 +24,66 @@ import (
...
+	for _, svcPort := range svc.Spec.Ports {
+		if (port.Type == intstr.Int && int32(svcPort.Port) == port.IntVal) || (port.Type == intstr.String && svcPort.Name == port.StrVal) {
+			return &svcPort, nil
+		}
+	}
+	return nil, errors.NewServiceUnavailable(fmt.Sprintf("no service port %q found for service %q", port.String(), svc.Name))
+}
+
 // ResourceLocation returns a URL to which one can send traffic for the specified service.
 func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string) (*url.URL, error) {
```
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: You have change the meaning of the id field (it used to be scheme : service name : port) it is now just the service name. I believe the caller at either https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L153-L155 or https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go#L116 will need to be fixed to reflect this.

Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: I did. But I am struggling to find out where we decide/document that a service name (which comes from APIService as far as I see) can be in this colon syntax.

**Comment by deads2k** [comment]
> You have change the meaning of the id field (it used to be scheme : service name : port) it is now just the service name. I believe the caller at either https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L153-L155 or https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go#L116 will need to be fixed to reflect this.

The service name has always been intended to be just a name, all the way back to the initial comment from @lavalamp about doing so here back in november: https://github.com/kubernetes/kubernetes/pull/37561#discussion_r90325567 .

If it changed, then it was an accidental quirk of implementation that change the API from its intended and documented shape.  If we want a port, that port name or int should be structured in the API as such and not glommed as a string.  And I was strongly swayed by Daniel's point in the initial version of the API.

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go (hunk 3 above)
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: Ran your fix on my GCE cluster. Despite my worries looks like it works just fine.

**Comment by sttts** [comment]
@k8s-bot pull-kubernetes-node-e2e test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-e2e-kops-aws test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Pull request merged**